packages/back-end/src/routers/project/project.controller.ts

```diff
@@ -31,11 +37,35 @@ export const postProject = async (
 ) => {
   req.checkPermissions("manageProjects");
 
+  const datasources: string[] = req.body.datasources || [];
+  const metrics: string[] = req.body.metrics || [];
   const { name } = req.body;
   const { org } = getOrgFromReq(req);
 
+  for (let i = 0; i < datasources.length; i++) {
+    const datasource = await getDataSourceById(datasources[i], org.id);
+    if (!datasource) {
+      res.status(403).json({
+        message: "Invalid datasource: " + datasources[i],
+      });
+      return;
+    }
+  }
+
+  for (let i = 0; i < metrics.length; i++) {
+    const metric = await getMetricById(metrics[i], org.id);
+    if (!metric) {
+      res.status(403).json({
+        message: "Invalid metric: " + metrics[i],
+      });
+      return;
+    }
+  }
```

Comment:
For the above, I think it might be worthwhile to create a `getDataSourcesByIds` and `getMetricsByIds` that both accept a list of IDs to minimize roundtrips to the DB

Reply:
Good call! added